### ISSUE TITLE
Performance notes: Qwen3.8-Flash-Next (qwen4_exp) single-stream decode on Apple Silicon — measured boundaries (mx.compile, MTP economics, load sensitivity)

### ISSUE THREAD

**Issue by moc375** [jundot/omlx#3353]
Sharing a set of measured performance findings for qwen4_exp on M3 Ultra (512GB, Flash-Next oQ4e, omlx 0.6.4, MTP depth-5) — including two falsified optimization paths. The negative results are the useful part: they document **where the existing serving paths are already optimal**, which saves the next person the month we spent finding out.

## Baselines (quiet GPU, single stream, engine-timed, ≥5 runs)

| workload | tok/s |
|---|---|
| prose, MTP depth-5 | 52.4 (46.8-57.7) |
| tool-calling / structured output, MTP | ~59 (accept 87-92%, tok/cycle 3.1) |
| greedy (MTP off) | 29.7 — **load-insensitive** |
| long-context decode | 41 @15k → 31.9 @91k (~20% droop over 6× context; GDN+QSA doing exactly what it was designed to do) |
| prefill, cache-miss | ~808 tok/s flat 15k-91k |

## Finding 1: whole-graph `mx.compile` cannot beat the eager B=1-class paths — they are already internally compiled

Two controlled attempts, both losses:

**(a) greedy (1,1) through a compiled decode lane**: 17.3 vs eager 29.6 tok/s (-42%). The eager single-token path runs the fused B1/T1 GDN decode prework and ~97-100 pre-compiled inner decode paths; a whole-graph compile swaps those specialized kernels for generic ones and loses more than the dispatch it saves.

**(b) MTP verify (1,T) through a compiled verify graph** — the careful version: bit-exact T>1 graph (per-position rope, multi-block pooled-indexer completion, per-position causal admission, all-position logits), zero-copy functional-leaf rollback, and **tracing under the eager runtime's own kernel branches** (`target_verify=True` threaded through every module call + the class-level QuantizedLinear qmm routing armed at trace time — the CUDA-graph-equivalence move). Result: 46-48ms/cycle vs eager 43ms; 37.8 vs 52.4 tok/s. All T=1..6 engagements verified by per-shape logging; acceptance within baseline (numerics gate passed).

**The boundary**: `mx.compile` wins exactly where the eager path is dispatch-heavy AND unspecialized — multi-row continuous batching (our PR #3334: host dispatch 8.8→2.0ms, step -18% at B=4, in production soak since). At B=1/T≤6 the eager paths already carry upstream's internal compilation, so a whole-graph compile is re-compiling an optimized path. We could not find a dispatch residue to save.

## Finding 2: MTP is GPU-throughput-bound, greedy is latency-bound — sibling-load sensitivity differs 40% vs 0%

| path | quiet | with sibling instance generating |
|---|---|---|
| MTP depth-5 | 52.4 | 31.6 (**-40%**) |
| greedy | 29.7 | 29.6 (±0) |

Methodology implication: **comparing a GPU-bound path against a CPU/latency-bound path under one load condition is structurally misleading** (our own MTP-vs-greedy verdict flipped once we measured both conditions). Any MTP benchmark on shared hardware should record sibling-instance load in the measurement window, from per-instance logs, not cumulative averages.

## Finding 3: the adaptive depth controller is near-optimal

Fixed-depth sweep (instrument: PR #3352): adaptive 52.4 vs pinned-5 55.5 (+6% max). Per-position conditional acceptance decays 0.74→0.45 across depths 1-5, capping depth-5 EV at ~2.8 tok/cycle. The controller's content adaptation is worth more than the pin loses (tool-calling climbs to 3.1 tok/cycle on its own).

## Finding 4: draft cost and orchestration are not the MTP bottleneck

Cycle anatomy (quiet, prose): verify forward ~40ms (92%), draft ~3ms (6%), sample/cache/rollback <0.3ms. The earlier hypothesis of ~10ms/cycle orchestration tax did not survive accounting against the four engine timing buckets.

## What we concluded (and share to save others the time)

On this hardware the remaining single-stream headroom is in **incremental fusion on top of the specialized eager kernels** (fusing the small ops between the already-fused GDN/MoE kernels), not in graph-level compilation, depth tuning, or draft-side work. Cross-checking the ecosystem: MLX tg 65.3 / llama.cpp tg 40.1 for a 3B-active GDN-hybrid on the same chip class (llama.cpp GDN Metal had its own long march, issues #19480/#21840) — omlx's Flash-Next serving is at or above the ecosystem frontier for this model class.

All experiment code from (a)/(b) is env-gated and tested (bit-exact tiny-model suites); happy to contribute any piece if there's interest, but the falsifications above are the transferable part.

Credit where due: the losses in Finding 1 are losses *because* the upstream eager paths are this good already.

**Comment by moc375** [comment]
# Flash-Next Decode Optimization Path Map

> Two months of measured exploration on M3 Ultra + Qwen3.8-Flash-Next (125B, 6B active, oQ4e).
> Every verdict backed by same-window controlled A/B. This map saves you the month we spent.

## Visual Map (Mermaid)

```mermaid
graph TB
    GOAL["🎯 Goal: 100+ tok/s<br/>single-stream"]

    %% ═══ WINS ═══
    subgraph WINS["✅ VERIFIED WINS (Production)"]
        MTP["MTP Depth-5<br/>52.4 tok/s quiet<br/>(vs 29.7 greedy)"]
        TOOL["Tool-Call MTP<br/>87-92% accept<br/>~59 tok/s"]
        CMP_B2["mx.compile B≥2<br/>dispatch -77%<br/>step -18%<br/>→ PR #3334"]
        CACHE["Prefix Cache<br/>77-83% hit rate<br/>TTFT 77-220ms @45K"]
        PREHEAT["Cache Preheat<br/>real content replay<br/>dual-instance"]
        ISO["GPU Isolation<br/>quiet 52.4<br/>loaded 31.6 (-40%)"]
    end

    %% ═══ FALSIFIED ═══
    subgraph DEAD["❌ FALSIFIED (Data-Closed)"]
        CMP_B1["mx.compile B=1<br/>17.3 vs 29.6 (-42%)<br/>eager B1/T1 fused wins"]
        CMP_VK["mx.compile Verify T=k<br/>37.8 vs 52.4<br/>eager verify pre-fused"]
        BATCH_MTP["Batched MTP B≥2<br/>verify scales rows×pos<br/>ceiling 60 tok/s"]
        QSA_SPARSE["QSA Sparse Kernel<br/>dense wins <30K ctx<br/>custom 88ms vs dense 64ms"]
        MOE_KERNEL["MoE Custom Kernel<br/>11% of step time<br/>custom loses"]
        QUANT["Quantization<br/>latency-bound<br/>15% BW util"]
    end

    %% ═══ ROOT CAUSE ═══
    LAW["⚖️ THE LAW<br/>omlx eager B=1 paths already<br/>internally compiled (97-100 paths)<br/>mx.compile wins ONLY where<br/>eager is unspecialized + dispatch-heavy<br/>= B≥2 continuous batching, nowhere else"]

    %% ═══ FRONTIER ═══
    subgraph OPEN["⏳ OPEN FRONTIER"]
        FUSION["GDN Projection Fusion<br/>month-level Metal kernel<br/>target 70-85 tok/s<br/>ONLY remaining path to 100+"]
        MIRROR["Cross-Instance Mirror<br/>session real-time sync<br/>failover with zero TTFT cost"]
    end

    %% ═══ EDGES ═══
    GOAL --> MTP
    GOAL --> CMP_B2
    GOAL --> CACHE
    GOAL --> FUSION

    MTP --> TOOL
    MTP -->|"B≥2?"| BATCH_MTP

    CMP_B2 -->|"B=1?"| CMP_B1
    CMP_B1 --> LAW
    CMP_B1 --> CMP_VK

    CACHE --> PREHEAT
    CACHE --> MIRROR

    FUSION -->|"tried sparse"| QSA_SPARSE
    QSA_SPARSE --> MOE_KERNEL
    QSA_SPARSE --> QUANT

    style GOAL fill:#ff9800,color:#fff
    style WINS fill:#1b5e20,color:#e8f5e9
    style DEAD fill:#b71c1c,color:#ffcdd2
    style LAW fill:#4a148c,color:#e1bee7
    style OPEN fill:#e65100,color:#fff3e0
    style MTP fill:#2e7d32,color:#fff
    style TOOL fill:#2e7d32,color:#fff
    style CMP_B2 fill:#2e7d32,color:#fff
    style CACHE fill:#2e7d32,color:#fff
    style PREHEAT fill:#2e7d32,color:#fff
    style ISO fill:#2e7d32,color:#fff
    style FUSION fill:#ef6c00,color:#fff
    style MIRROR fill:#ef6c00,color:#fff
    style LAW fill:#6a1b9a,color:#fff
```

## Summary Table

| Path | Verdict | Measured | Key Insight |
|---|---|---|---|
| MTP ON vs OFF | ✅ WIN | 52.4 vs 29.7 quiet | +76% on quiet GPU |
| MTP Depth-5 | ✅ OPTIMAL | 52.4 (d1=47.2, d8=flat) | acceptance decays by position |
| MTP Tool-Call | ✅ STRONGEST | ~59 tok/s, 87-92% accept | agent workload is MTP's best case |
| Depth Controller | ✅ NEAR-OPTIMAL | +6% max from forced-5 | adaptive ≈ optimal |
| mx.compile B≥2 | ✅ WIN (PR #3334) | dispatch -77%, step -18% | **only where eager unspecialized** |
| mx.compile B=1 | ❌ FALSIFIED | 17.3 vs 29.6 (-42%) | eager has B1/T1 fused GDN |
| mx.compile Verify | ❌ FALSIFIED | 37.8 vs 52.4 | eager verify pre-fused |
| Batched MTP B≥2 | ❌ FALSIFIED | ceiling 60 tok/s | verify scales rows×positions |
| QSA Sparse Kernel | ❌ FALSIFIED | custom 88ms vs dense 64ms | dense wins <30K context |
| MoE Custom Kernel | ❌ NOT WORTH | 11% of step time | not the bottleneck |
| Quantization | ❌ NO EFFECT | 15% BW utilization | latency-bound, not bandwidth |
| DFlash drafter | ❌ N/A | no model for Flash-Next | same verify wall |
| DSpark drafter | ❌ N/A | NVIDIA-only | same verify wall |
| Exact Hybrid Proj | ❌ NOT A GAP | shape mismatch | correct skip |
| Draft indexer reuse | ⬇️ LOW ROI | draft only 3ms/cycle (6%) | not the lever |
| GDN Fusion | ⏳ OPEN | target 70-85 | month-level Metal kernel |
| Cross-Instance Mirror | ⏳ OPEN | TTFT→0 on failover | future work |

## The Transferable Law

```
┌─────────────────────────────────────────────────────────────────────┐
│  omlx's eager single-stream (B=1) decode paths are ALREADY          │
│  internally compiled by upstream:                                    │
│    • 97-100 pre-compiled hyper-connection decode paths              │
│    • B1/T1 fused GDN decode prework                                 │
│    • verify-shape qmm/GEMV kernels                                  │
│    • fused verify attention patches                                 │
│                                                                       │
│  mx.compile is re-compiling an already-optimized path —             │
│  there is no dispatch residue to save at B=1.                        │
│                                                                       │
│  mx.compile WINS exactly where eager is:                             │
│    (a) dispatch-heavy  →  B≥2 continuous batching                   │
│    (b) unspecialized   →  no pre-compiled inner paths               │
│  Both conditions must hold simultaneously.                          │
└─────────────────────────────────────────────────────────────────────┘
```

## Load Sensitivity (Benchmark Methodology)

| Path | Quiet GPU | Loaded GPU | Sensitivity |
|---|---|---|---|
| MTP depth-5 | 52.4 tok/s | 31.6 tok/s | **-40%** (GPU-throughput-bound) |
| Greedy | 29.7 tok/s | 29.6 tok/s | **±0** (CPU/latency-bound) |

**Implication**: comparing a GPU-bound path vs a CPU-bound path under one load condition
is structurally misleading — both conditions must be measured.

**Comment by moc375** [comment]
## Optimization Path Map (Mermaid, ASCII-safe)

> Replacing the previous comment which failed to render due to Unicode in subgraph labels.

```mermaid
graph TB
    GOAL["Goal: 100+ tok/s single-stream"]

    subgraph wins["VERIFIED WINS"]
        MTP["MTP Depth-5: 52.4 tok/s quiet vs 29.7 greedy"]
        TOOL["Tool-Call MTP: 87-92% accept, 59 tok/s"]
        CMP2["mx.compile B2+: dispatch -77%"]
        CACHE["Prefix Cache: 77-83% hit, TTFT 77ms"]
        PREHEAT["Cache Preheat: dual-instance"]
        ISO["GPU Isolation: quiet 52.4 vs loaded 31.6"]
    end

    subgraph dead["FALSIFIED - DATA CLOSED"]
        B1["Compile B=1: 17.3 vs 29.6"]
        VK["Compile Verify T=k: 37.8 vs 52.4"]
        BMTP["Batched MTP B2+: ceiling 60 t/s"]
        QSA["QSA Sparse: dense wins under 30K"]
        QUANT["Quantization: latency-bound"]
    end

    LAW["THE LAW: omlx eager B=1 already compiled. mx.compile wins ONLY at B2+"]

    subgraph open["OPEN FRONTIER"]
        FUSION["GDN Fusion: month-level, 70-85 t/s"]
        MIRROR["Cross-Instance Mirror: future"]
    end

    GOAL --> MTP
    GOAL --> CMP2
    GOAL --> CACHE
    GOAL --> FUSION
    MTP --> TOOL
    MTP -->|B2+?| BMTP
    CMP2 -->|B=1?| B1
    B1 --> LAW
    B1 --> VK
    CACHE --> PREHEAT
    CACHE --> MIRROR
    FUSION -->|tried sparse| QSA
    QSA --> QUANT
```
